README.md
# Imaginex

Pre-compiled Pythong lambda based on Pillow.
Shares API with NextJS and is mainly meant as replacement of shitty build-in NextJS image optimizer.

However, can be plugged into any other project with straight forward API.

Currently supported image formats are
PNG, JPEG, ICO, WEBP, GIF, TIFF, BMP and others based on the former formats supported by pillow package (e.g. -
JPEG2000, JFIF, MPO, ...)

## Usage

- Import/download zips from this package,
- Use them as lambda's code,
- Allow lambda to access S3 bucket with images and set S3_BUCKET_NAME env variable,
- Call lambda via API Gateway with query params:
    - url (relative to bucket or absolute to anywhere)
    - w (width for resizing) or h (height for resizing)
    - q (quality for optimizing)

Response is base64 encoded image.

Supports all formats supported by Pillow.
Faster and easily deployable than NextJS image optimizer.

## Notes

Only 3.8 python runtime is supported!

If you want to compile python modules for Lambda, make sure to set `--platform manylinux2014_x86_64`. It's important to
set this in order to have compatible binaries in Lambda's execution environment.

---

Child directory of: http://github.com/sladg/nextjs-lambda


CHANGELOG.md
# Changelog


## [v0.18.2](https://github.com/sladg/imaginex-lambda/compare/v0.18.1...v0.18.2)

* 🐛 fix(s3): correct order of ifs, fixed parsing edgecases [[ddcf30b38a4386abb6ed2655d925af169bca2cff](https://github.com/sladg/imaginex-lambda/commit/ddcf30b38a4386abb6ed2655d925af169bca2cff))]


## [v0.18.1](https://github.com/sladg/imaginex-lambda/compare/v0.18.0...v0.18.1)

* 🐛 fix(s3): parse encoded http uri [[dd6596ad37d1f100c1ec1e2951701384452b67a6](https://github.com/sladg/imaginex-lambda/commit/dd6596ad37d1f100c1ec1e2951701384452b67a6))]


## [v0.18.0](https://github.com/sladg/imaginex-lambda/compare/v0.17.0...v0.18.0)

* ✨ feat(s3): allow for specifying bucketname withing the url [[8b8d57e18952c2c564203ba45538329845be2951](https://github.com/sladg/imaginex-lambda/commit/8b8d57e18952c2c564203ba45538329845be2951))]


## [v0.17.0](https://github.com/sladg/imaginex-lambda/compare/v0.7.0...v0.17.0)

* feat(img_lib): portrait vs landscape [[6a09c9472dc86578d8b08b3921f5749dc3ece9b4](https://github.com/sladg/imaginex-lambda/commit/6a09c9472dc86578d8b08b3921f5749dc3ece9b4))]
* fix(img_lib): logger fix [[7cc415318916767fafba6d5738453bbeae6d11c4](https://github.com/sladg/imaginex-lambda/commit/7cc415318916767fafba6d5738453bbeae6d11c4))]
* test(formats): moved list of not supported formats into fun [[7a2011c885c0323100470bfdaa3cb1e26d32dc9f](https://github.com/sladg/imaginex-lambda/commit/7a2011c885c0323100470bfdaa3cb1e26d32dc9f))]
* refactor(tests,libs): added tests, readme, docstrings [[073f9ff91ecf3f51ef9ad289fcb880dbaac1c6fd](https://github.com/sladg/imaginex-lambda/commit/073f9ff91ecf3f51ef9ad289fcb880dbaac1c6fd))]
* test(success): magicmock [[f5dc6b14e37615f8db397fd9091dfb1612cc95c0](https://github.com/sladg/imaginex-lambda/commit/f5dc6b14e37615f8db397fd9091dfb1612cc95c0))]
* test(success): height [[b5b8c96c76b11eeac5f91e643ddd18717e587d13](https://github.com/sladg/imaginex-lambda/commit/b5b8c96c76b11eeac5f91e643ddd18717e587d13))]
* test(tests): updated tests [[6453d0e294c7bfde91af8ed8f52444e477be2ab9](https://github.com/sladg/imaginex-lambda/commit/6453d0e294c7bfde91af8ed8f52444e477be2ab9))]
* fix(img_lib): missing import [[78d084d301d0b8763cb2c73ef317caf83564e133](https://github.com/sladg/imaginex-lambda/commit/78d084d301d0b8763cb2c73ef317caf83564e133))]
* refactor(utils): logger [[5c0bf36a33f7389c8f699d8e9f2822f14cb88239](https://github.com/sladg/imaginex-lambda/commit/5c0bf36a33f7389c8f699d8e9f2822f14cb88239))]
* refactor(utils): typings [[4ff2390ba37c029eae656526b4074900c69a76b8](https://github.com/sladg/imaginex-lambda/commit/4ff2390ba37c029eae656526b4074900c69a76b8))]
* fix(exceptions): missing import [[02d01d0189dc4a15746ff7ebdd803eed48d2a8c5](https://github.com/sladg/imaginex-lambda/commit/02d01d0189dc4a15746ff7ebdd803eed48d2a8c5))]
* refactor(lib): restructuring files structure [[1f71342ddd4414ddd99c1daeb3f6ff4635d272df](https://github.com/sladg/imaginex-lambda/commit/1f71342ddd4414ddd99c1daeb3f6ff4635d272df))]
* refactor(handler,-utests): updates, refactoring, testing [[87d9e0611f7b051808dc877b388936bfc80c93d8](https://github.com/sladg/imaginex-lambda/commit/87d9e0611f7b051808dc877b388936bfc80c93d8))]


## [v0.7.0](https://github.com/sladg/imaginex-lambda/compare/v0.5.0...v0.7.0)

* 🐛 fix(ratio): prevent future "division by zero" bugs on ratio calculation [[ef778bc8a147701265e162fc8d93089d9898ff1f](https://github.com/sladg/imaginex-lambda/commit/ef778bc8a147701265e162fc8d93089d9898ff1f))]
* 🐛 fix(ratio): handle very small image edge case [[563f886fe3a31a14e22c473b98b2bcd6d26c9759](https://github.com/sladg/imaginex-lambda/commit/563f886fe3a31a14e22c473b98b2bcd6d26c9759))]
* ♻️ refactor(handler): isolates the handler code from the image processing [[6731100aa4cb71acbb6aba48755717339987b80a](https://github.com/sladg/imaginex-lambda/commit/6731100aa4cb71acbb6aba48755717339987b80a))]
* 🧪 test(handler): replaces example JSON files for a programmatic approach [[36f2fe9659bd0781dc2be2284089b455addf41bf](https://github.com/sladg/imaginex-lambda/commit/36f2fe9659bd0781dc2be2284089b455addf41bf))]


## [v0.5.0](https://github.com/sladg/imaginex-lambda/compare/v0.2.6...v0.5.0)

* 📦 ci(tests): include test command before release, correct folder naming [[b609483a5bbcaaa00c97fae8843502970529456b](https://github.com/sladg/imaginex-lambda/commit/b609483a5bbcaaa00c97fae8843502970529456b))]
* 💎 style(pyproject): improved style, added easy command for tests [[c977063c26a9b641843e55e6984d59ebab0bc1b5](https://github.com/sladg/imaginex-lambda/commit/c977063c26a9b641843e55e6984d59ebab0bc1b5))]
* Merge pull request #1 from fabiob/optimizations [[e6236e4d2a99887d5477602c11fd59e99624c532](https://github.com/sladg/imaginex-lambda/commit/e6236e4d2a99887d5477602c11fd59e99624c532))]
* 📈 perf(handler): PEP8, memory usage, dependencies, tests [[c41d267efc584797fbaa37e24bef42bca3255728](https://github.com/sladg/imaginex-lambda/commit/c41d267efc584797fbaa37e24bef42bca3255728))]


## [v0.2.6](https://github.com/sladg/imaginex-lambda/compare/v0.2.4...v0.2.6)

* 🐛 fix(release): change package to correct one [[cd762f6b8a79e55f77b8520775602e10380f53b0](https://github.com/sladg/imaginex-lambda/commit/cd762f6b8a79e55f77b8520775602e10380f53b0))]
* 🐛 fix(exports): correctly export version and name of the package [[e7b74f14d158ab7c00cc930c09e8770b903aa81d](https://github.com/sladg/imaginex-lambda/commit/e7b74f14d158ab7c00cc930c09e8770b903aa81d))]


## [v0.2.4](https://github.com/sladg/imaginex-lambda/compare/v0.2.3...v0.2.4)

* fix(output): correct naming of zips [[c60f6e7b05eb4c37a9823181766564a495051d6a](https://github.com/sladg/imaginex-lambda/commit/c60f6e7b05eb4c37a9823181766564a495051d6a))]


## [v0.2.3](https://github.com/sladg/imaginex-lambda/compare/v0.2.2...v0.2.3)

* fix(binaries): correctly compile libraries so they are compatible with lambda env [[88250e70ce20582cdb37f97a5f7780718c4ad456](https://github.com/sladg/imaginex-lambda/commit/88250e70ce20582cdb37f97a5f7780718c4ad456))]



test/test_formats.py
import base64
from io import BytesIO
from tempfile import TemporaryFile
from unittest.mock import patch
from PIL import Image

from imaginex_lambda.handler import download_and_optimize, handler

import pytest
from unittest.mock import MagicMock


@pytest.mark.parametrize('img_type,expected_type', [
    ('PNG', 'image/png'),
    ('JPEG', 'image/jpeg'),
    ('JPEG2000', 'image/jpx'),
    ('ICO', 'image/x-icon'),
    ('MPO', 'image/jpeg'),
    ('WEBP', 'image/webp'),
    ('GIF', 'image/gif'),
    ('TIFF', 'image/tiff'),
    ('BMP', 'image/bmp'),
])
def test_download_and_optimize_formats(img_type, expected_type):
    # List of possibly available formats in the future (supported only by PIL atm)
    # ('DIB', 'image/dib'),        # HandlerError: Unsupported image format
    # ('PPM', 'image/ppm'),        # HandlerError: Unsupported image format
    # ('BUFR', 'image/BUFR'),      # HandlerError: Unsupported image format
    # ('PCX', 'image/PCX'),        # HandlerError: Unsupported image format
    # ('DDS', 'image/DDS'),        # HandlerError: Unsupported image format
    # ('ICNS', 'image/ICNS'),      # HandlerError: Unsupported image format
    # ('IM', 'image/IM'),          # HandlerError: Unsupported image format
    # ('SGI', 'image/SGI'),        # HandlerError: Unsupported image format
    # ('SPIDER', 'image/SPIDER'),  # HandlerError: Unsupported image format
    # ('TGA', 'image/TGA'),        # HandlerError: Unsupported image format
    # ('BLP', 'image/blp'),        # ValueError: Unsupported BLP image mode
    # ('FITS', 'image/FITS'),      # OSError: FITS save handler not installed
    # ('GRIB', 'image/GRIB'),      # OSError: GRIB save handler not installed
    # ('HDF5', 'image/HDF5'),      # OSError: HDF5 save handler not installed
    # ('WMF', 'image/WMF'),        # OSError: WMF save handler not installed
    # ('MSP', 'image/MSP'),        # OSError: cannot write mode RGB as Palm
    # ('PALM', 'image/PALM'),      # OSError: cannot write mode RGB as Palm
    # ('XBM', 'image/XBM'),        # OSError: cannot write mode RGB as XBM
    # ('PDF', 'image/PDF'),        # PIL.UnidentifiedImageError
    # ('EPS', 'image/EPS'),        # Resize error
    # arrange
    original_w, original_h, new_q, new_w = 300, 300, 80, 256

    sample_url = "https://example.com"
    img = Image.new('RGB', (original_w, original_h), color=(255, 0, 0))
    tmp_img = TemporaryFile()
    img.save(tmp_img, format=img_type)
    download_image_mock = MagicMock(return_value=(tmp_img, {}))

    lambda_context = {
        'queryStringParameters':
            {
                'q': new_q, 'w': new_w, 'url': sample_url
            }
    }

    # act
    with patch('imaginex_lambda.lib.img_lib.download_image', download_image_mock):
        opt_img_bytes, content_type, ratio = download_and_optimize(sample_url, new_q, new_w, None, '')
        res = handler(lambda_context, None)
    tmp_img.close()

    # assert
    opt_img_res = Image.open(BytesIO(base64.b64decode(res['body'].encode())))
    assert content_type == expected_type
    assert opt_img_res.width == new_w


test/test_success.py
from unittest.mock import patch

import pytest

from imaginex_lambda.handler import handler, S3_BUCKET_NAME, download_and_optimize
from unittest.mock import MagicMock


def test_handler_success():
    """
    This test will mock the result of the 'download_and_optimize' function
    to return a fixed value. This way we can check if the handler single responsibility
    (to parse the context and generate a formatted response) is working correctly.

    Specific image processing tests should call 'download_and_optimize' directly,
    so it is easier to check for exceptions and edge cases.
    """

    context = {'queryStringParameters': {'url': 'abc.png', 'q': '50', 'w': '100'}}
    fake_optimization_return = MagicMock(return_value=((b'abcdef', 'application/someimage', 0.3)))

    with patch('imaginex_lambda.handler.download_and_optimize', fake_optimization_return) as p:
        r = handler(context, None)

    assert r['statusCode'] == 200
    assert r['isBase64Encoded'] is True
    assert r['headers']['Content-Type'] == 'application/someimage'
    assert r['headers']['X-Optimization-Ratio'] == '0.3000'


@pytest.mark.parametrize('expected_ratio,expected_type,q,w,url', [
    (0.0037, 'image/png', 80, 100, "example.png"),
    (0.0109, 'image/jpeg', 40, 250, "https://s3.eu-central-1.amazonaws.com/fllite-dev-main/"
                                    "business_case_custom_images/sun_valley_2_5f84953fef8c6_63a2668275433.jfif"),
    (0.1770, 'image/jpeg', 80, 100, "http://site.meishij.net/r/58/25/3568808/a3568808_142682562777944.jpg"),
    (0.6733, 'image/jpeg', 80, 100, "https://www.gravatar.com/avatar/617bdc1719f77448a4f96eb92e1ac02b?s=80&d=mp"),
])
def test_process_success(expected_ratio, expected_type, q, w, url):
    if not url.startswith('http') and S3_BUCKET_NAME is None:
        pytest.skip('specify a value for S3_BUCKET_NAME to run S3 tests')

    image_data, content_type, ratio = download_and_optimize(url=url, quality=q, width=w, height=None, bucket_name='')
    assert isinstance(image_data, bytes)
    assert content_type == expected_type
    assert round(ratio, 4) == expected_ratio


@pytest.mark.parametrize('w,h,q', [
    (None, 100, 80),
    (80, None, 80),
    (None, 200, 60),
    (200, None, 60),
])
@pytest.mark.parametrize('expected_type,url', [
    ('image/jpeg', "https://s3.eu-central-1.amazonaws.com/fllite-dev-main/"
                   "business_case_custom_images/sun_valley_2_5f84953fef8c6_63a2668275433.jfif"),
    ('image/jpeg', "https://www.gravatar.com/avatar/617bdc1719f77448a4f96eb92e1ac02b?s=80&d=mp"),
])
def test_process_success(expected_type, url, w, h, q):
    if not url.startswith('http') and S3_BUCKET_NAME is None:
        pytest.skip('specify a value for S3_BUCKET_NAME to run S3 tests')
    image_data, content_type, ratio = download_and_optimize(url=url, quality=q, width=w, height=h, bucket_name='')
    assert isinstance(image_data, bytes)
    assert content_type == expected_type


test/test_failures.py
from unittest.mock import patch

import pytest

from imaginex_lambda.handler import handler, download_and_optimize
from imaginex_lambda.lib.exceptions import HandlerError


def test_no_url():
    r = handler({'queryStringParameters': {}}, None)
    assert r['statusCode'] == 422
    assert r['headers']['Content-Type'] == 'application/json'
    assert "url is required" in r['body']


@pytest.mark.parametrize('width', [None, -1, -100, 0])
def test_invalid_width(width):
    r = handler({'queryStringParameters': {'url': 'abc.png', 'width': width}}, None)
    assert r['statusCode'] == 422
    assert r['headers']['Content-Type'] == 'application/json'
    assert "Width or height must be defined" in r['body']


def test_s3_not_configured():
    with patch('imaginex_lambda.handler.S3_BUCKET_NAME', None):
        r = handler({'queryStringParameters': {'url': 'abc.png', 'w': 50}}, None)
        assert r['statusCode'] == 500
        assert r['headers']['Content-Type'] == 'application/json'
        assert "must specify a value for S3_BUCKET_NAME" in r['body']


def test_s3_invalid_bucket():
    with patch('imaginex_lambda.handler.S3_BUCKET_NAME', 'i'):
        r = handler({'queryStringParameters': {'url': 'abc.png', 'w': 50}}, None)
        assert r['statusCode'] == 500
        assert r['headers']['Content-Type'] == 'application/json'
        assert "InvalidBucketName" in r['body']


def test_download_and_optimize_with_invalid_url():
    # arrange
    url = ''
    quality = 50
    width = 100

    # act and assert
    with pytest.raises(HandlerError):
        download_and_optimize(url, quality, width, None, '')


def test_download_and_optimize_with_zero_width():
    # arrange
    url = 'https://example.com/image.png'
    quality = 50
    width = 0

    # act and assert
    with pytest.raises(HandlerError):
        download_and_optimize(url, quality, width, None, '')


imaginex_lambda/lib/utils.py
import base64
from typing import IO, Dict, Any
from urllib.parse import urlparse, unquote

import filetype
import logging

from PIL.Image import Image

from imaginex_lambda.lib.exceptions import HandlerError

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')


def success(image_data: bytes, headers: Dict[str, Any]) -> Dict[str, Any]:
    return {
        'statusCode': 200,
        'body': base64.b64encode(image_data).decode(),
        'isBase64Encoded': True,
        'headers': headers
    }

def is_absolute(url: str) -> bool:
    return bool(urlparse(url).netloc)

def is_s3(url: str) -> bool:
    return unquote(url).startswith('s3://')

def get_extension(buffer: IO[bytes]) -> Dict:
    """
    Determines the file extension and content type of an image file contained in the given buffer.

    Args:
        buffer (IO[bytes]): The buffer containing the image file.

    Returns:
        Dict[str, str]: A dictionary containing the content_type and extension of the image file.

    Raises:
        HandlerError: If the image format is not supported.
    """
    logger.info("Getting extension...")

    kind = filetype.guess(buffer)
    if kind is None:
        raise HandlerError('Unsupported image format')
    content_type = kind.mime
    extension = content_type.upper().replace('IMAGE/', '')
    if extension == 'JPX':
        extension = 'JPEG2000'
    elif extension == 'X-ICON':
        extension = 'ICO'
    logger.info(f"Extension: {extension}")
    return {'content_type': content_type, 'extension': extension}


def cast_to_int(value):
    try:
        return int(value)
    except (ValueError, TypeError):
        return None


def is_landscape(image: Image) -> bool:
    """
    Checks whether an image is landscape or portrait.

    Args:
        image (PIL.Image): The opened PIL Image object.

    Returns:
        bool: True if the image is landscape, False if it is portrait.
    """
    # Get the width and height of the image
    width, height = image.size

    # Compare the width and height to determine if the image is landscape or portrait
    if width > height:
        return True
    else:
        return False


imaginex_lambda/lib/img_lib.py
import os
import shutil
from contextlib import ExitStack
from io import BytesIO
from tempfile import TemporaryFile
from typing import IO, Tuple, Dict, Any, Optional
from urllib.request import urlopen
from urllib.parse import unquote

import botocore.session
from PIL import Image

from imaginex_lambda.lib.exceptions import HandlerError
from imaginex_lambda.lib.utils import is_absolute, is_s3, get_extension, logger, is_landscape

# Pillow supported formats:
# BLP, BMP, DDS, DIB, EPS, GIF, ICNS, ICO, IM, JPG, JPEG, MSP, PCX, PNG, PPM, SPIDER, TGA, TIFF, WEBP, XBM

session = botocore.session.get_session()
s3_client = session.create_client('s3')


# @TODO: Add placeholder image for errors.

def download_image(buffer: IO[bytes], img_url: str, chunk_size: int = 1024) -> Tuple[IO[bytes], Dict[str, Any]]:
    """
    Function responsible for downloading an image file from a given URL and writing its contents to a buffer.
    It takes two arguments, buffer and img_url, and returns a dictionary containing information about the downloaded
    image.

    Args:
        buffer: (IO[bytes]) A file-like object that the downloaded image will be written to.
        img_url: (str) A string representing the URL of the image to be downloaded.
    Returns:
        Tuple[IO[bytes], Dict[str, Any]]: dictionary containing information about the downloaded image
    """
    logger.info("Downloading image from %s", img_url)

    with urlopen(img_url) as r:
        content_type = r.headers['content-type']
        content_size = int(r.headers['content-length'])

        shutil.copyfileobj(r, buffer, chunk_size)

    logger.info("Downloaded image from %s. Content type: %s, content size: %d", img_url, content_type, content_size)
    return buffer, {'content_type': content_type, 'content_size': content_size}


def get_s3_image(buffer: IO[bytes], bucket_name: str, key: str, chunk_size: int = 1024) \
        -> Tuple[IO[bytes], Dict[str, Any]]:
    """
    Function responsible for downloading an image file from an Amazon S3 bucket and writing its contents to a buffer.
    It takes two arguments, buffer and key, and returns a dictionary containing information about the downloaded image.

    Args:
        buffer (IO[bytes]): The buffer to write the downloaded image contents to.
        bucket_name (str): The name of the S3 bucket to download the image from.
        key (str): The key of the S3 object to download.
        chunk_size (int): The chunk size to use when downloading the image.

    Returns:
        Tuple[IO[bytes], Dict[str, Any]]: A tuple containing the buffer with the downloaded image contents and a dictionary
        with information about the downloaded image (content_type and content_size).

    Raises:
        Exception: If `bucket_name` is not specified.
    """

    if not bucket_name:
        raise Exception('must specify a value for S3_BUCKET_NAME for S3 support')

    logger.info("Downloading image from S3 with key: %s", key)
    r = s3_client.get_object(Bucket=bucket_name, Key=key)

    content_type = r['ContentType']
    content_size = r['ContentLength']

    with r['Body'] as fin:
        shutil.copyfileobj(fin, buffer, chunk_size)

    logger.info("Downloaded image from S3 with key: %s. Content type: %s, content size: %d", key, content_type,
                content_size)
    return buffer, {'content_type': content_type, 'content_size': content_size}


def optimize_image(buffer: IO[bytes],
                   ext: str,
                   quality: int,
                   width: Optional[int] = None,
                   height: Optional[int] = None) -> bytes:
    """
    The optimize_image function is designed to optimize an image that is passed in. It resizes the image
    to the given width (if necessary), compresses the image to reduce its size, and returns the optimized image data.

    Args:
        buffer (IO[bytes]): buffer containing the image data to be optimized.
        ext (str): the file extension of the image, which is used to specify the format when saving the optimized image.
        quality (int): the quality of the compressed image. A higher quality will result in a larger file size, while a lower quality
            will result in a smaller file size.
        width (Optional[int]): the maximum width of the image. If the image is wider than this value, it will be resized to fit within
            this width.
        height (Optional[int]): the maximum height of the image. If the image is wider than this value, it will be resized to fit within
            this height.
    Returns:
        bytes: Optimized image data
    """

    logger.info("Optimizing image...")

    with ExitStack() as stack:
        img = stack.enter_context(Image.open(buffer))

        if width is not None and height is not None:
            if is_landscape(img):
                logger.info("Image is in landscape orientation, using width")
                height = None
            else:
                logger.info("Image is in portrait orientation, using height")
                width = None

        if width and width < img.width:
            logger.info(f"Resizing image given width {width}px...")
            new_height = int(width * img.height / img.width)
            logger.info(f"New height: {new_height}px")
            img = stack.enter_context(img.resize((width, new_height)))
            logger.info(f"Resized image to width:{width}px and height: {new_height}px")
        elif height and height < img.height:
            logger.info(f"Resizing image to the given {height}px...")
            new_width = int(height * img.width / img.height)
            logger.info(f"New width: {new_width}px")
            img = stack.enter_context(img.resize((new_width, height)))
            logger.info(f"Resized image to width: {new_width}px and height: {height}px")
        tmp = stack.enter_context(BytesIO())
        img.save(tmp, quality=quality, optimize=True, format=ext)
        tmp.seek(0)

        logger.info("Optimized image!")
        return tmp.read()


def download_and_optimize(url: str,
                          quality: int,
                          width: Optional[int],
                          height: Optional[int],
                          bucket_name: str,
                          chunk_size: int = 1024) -> Tuple[bytes, str, float]:
    """
    This is the function responsible for coordinating the download and optimization of the images. It should
    not concern itself with any lambda-specific information.

    Args:
        url (str): The URL of the image to download.
        quality (int): The quality to optimize the image (0 to 100).
        width (Optional[int]): The width of the image to resize to.
        height (Optional[int]): The height of the image to resize to.
        bucket_name (str): The name of the S3 bucket to download the image from.
        chunk_size (int): The chunk size to use when downloading the image.

    Returns:
        Tuple[bytes, str, float]: A tuple containing the optimized image data, content type, and the compression ratio.

    Raises:
        HandlerError: If `url` is empty, `width` and `height` are both empty or both provided, or if `width`
        or `height` are less than or equal to zero.

    """

    if not url:
        raise HandlerError('url is required')

    if width is None and height is None:
        raise HandlerError('Width or height must be defined')

    if width is not None and width <= 0:
        raise HandlerError('width must be greater than zero')

    if height is not None and height <= 0:
        raise HandlerError('height must be greater than zero')

    with TemporaryFile() as buffer:
        if is_s3(url):
            dynamic_bucket, key = unquote(url).replace('s3://', '').split('/', 1)
            logger.info("Dynamic bucket: %s", dynamic_bucket)
            buffer, _ = get_s3_image(buffer, dynamic_bucket, key, chunk_size)
        elif is_absolute(url):
            buffer, _ = download_image(buffer, url, chunk_size)
        else:
            key = url.strip('/')
            buffer, _ = get_s3_image(buffer, bucket_name, key, chunk_size)

        buffer.flush()
        original = os.stat(buffer.name).st_size
        mime = get_extension(buffer)
        content_type = mime['content_type']
        extension = mime['extension']

        image_data = optimize_image(
            buffer,
            ext=extension,
            quality=quality,
            width=width,
            height=height
        )

    ratio = len(image_data) / original if original != 0 else 0

    logger.info("Returning image and metadata")
    return image_data, content_type, ratio


imaginex_lambda/handler.py
import os
from typing import Dict, Optional

from PIL import Image

from imaginex_lambda.lib.exceptions import HandlerError, error
from imaginex_lambda.lib.img_lib import download_and_optimize
from imaginex_lambda.lib.utils import success, logger, cast_to_int

# @TODO: Add placeholder image for errors.

DOWNLOAD_CHUNK_SIZE = int(os.getenv('DOWNLOAD_CHUNK_SIZE', 1024))
DEFAULT_QUALITY_PERC = 70
S3_BUCKET_NAME = os.getenv('S3_BUCKET_NAME', None)


def handler(event: Dict, context: Optional[Dict]):
    """
    Lambda function handler.

    Its sole responsibility is to parse the context and generate a formatted response, including error responses.
    Any image processing logic should be performed by other functions, to make unit testing easier.
    """
    try:
        logger.info("Lambda function started")

        qs = event['queryStringParameters']
        url = qs.get('url', None)
        width = cast_to_int(qs.get('w', None))
        height = cast_to_int(qs.get('h', None))
        quality = int(qs.get('q', DEFAULT_QUALITY_PERC))

        logger.info(f"url={url}, width={width}, height={height} quality={quality}")

        image_data, content_type, optimization_ratio = download_and_optimize(url,
                                                                             quality,
                                                                             width,
                                                                             height,
                                                                             S3_BUCKET_NAME,
                                                                             DOWNLOAD_CHUNK_SIZE)
        logger.info("Returning success response")
        return success(image_data, {
            'Vary': 'Accept',
            'Content-Type': content_type,
            'X-Optimization-Ratio': f'{optimization_ratio:.4f}',
        })
    except HandlerError as exc:
        return error(str(exc), code=exc.code)
    except Exception as exc:
        return error(str(exc), code=500)


if __name__ == '__main__':
    from io import BytesIO
    import base64

    logger.info("Running test...")
    test_context = {
        'queryStringParameters':
            {
                'q': '40', 'h': '250', 'w': '250',
                'url': 'https://s3.eu-central-1.amazonaws.com/fllite-dev-main/'
                       'business_case_custom_images/sun_valley_2_5f84953fef8c6_63a2668275433.jfif'}
    }
    res = handler(test_context, None)
    img = res['body']
    Image.open(BytesIO(base64.b64decode(img.encode()))).show()
    logger.info("Test ended...")
